docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,71 @@
+
+This document is a draft of changes to ECMA-335 for the "runtime async" feature. When the feature is officially supported, it can be merged into the final ECMA-335 augments document.
+
+# Runtime-async
+
+Async is currently a feature implemented by various .NET languages as a compiler rewrite to support methods that can "yield" control back to their caller at specific "suspension" points. While effective, it's believed that implementation directly in the .NET runtime can provide improvements, especially in performance.
+
+## Spec modifications
+
+These are proposed modifications to the ECMA-335 specification for runtime-async.
+
+### I.8.4.5 Sync and Async Methods
+
+Methods may be either 'sync' or 'async'. Async methods have a special signature encoding, described in [### I.8.6.1.5 Method signatures].
+
+Sync methods are all other methods.
+
+Unlike sync methods, async methods support suspension. Suspension allows async methods to yield control flow back to their caller at certain well-defined suspension points, and resume execution of the remaining method at a later time or location, potentially on another thread.
+
+Async methods support the following suspension points:
+
+* Calling another async method. No special instructions need to be provided. If the callee suspends, the caller will suspend as well.
+* Using a library in the .NET corelib to "await" an "INotifyCompletion" type. The signatures of these methods shall be:
+  ```C#
+  // public static async2 Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  // public static async2 Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  ```
+
+Each of the above methods will have semantics analogous to the current AsyncTaskMethodBuilder.AwaitOnCompleted/AwaitUnsafeOnCompleted methods. After calling this method, in can be presumed that the task has completed.
+
+Async methods have the following restrictions:
+* Usage of the `localloc` and instruction is forbidden
+* The `lodloca` and `ldarga` instructions are redefined to return managed pointers instead of pointers.
```

Comment:
Note that `ldloca` and `ldarga` already return managed pointers in the spec. This should rather be a modification around the note we have on transient pointers here: https://github.com/dotnet/runtime/blob/main/docs/design/specs/Ecma-335-Augments.md#transient-pointers

There is a question of what level of behavior we want to specify. We can either specify
> Arguments / local variables of async methods are stored in unmanaged memory with an address that is guaranteed to be static in between suspension points, but that may change across suspension points.

or

> Arguments / local variables of async methods are stored in managed memory.

The latter precludes code like `unsafe { <unsafe code taking addresses of locals without any suspension points> }` in C#, but Roslyn already warns on that today.

I think in either case we also need to document that values of managed pointers and structs containing managed pointers are not preserved across suspension points (at least for an initial version).

Reply:
I picked the last one because it's the most restrictive. We can relax it if necessary.